virtcontainers/acrn_arch_base.go

```diff
@@ -0,0 +1,762 @@
+
+	acrnParams = append(acrnParams, "-s")
+	acrnParams = append(acrnParams, fmt.Sprintf("%d,virtio-blk,%s",
+		slot, blkdev.FilePath))
+
+	// Update the global array (BlkIndex<->slot)
+	// Used to identify slots for the hot-plugged virtio-blk devices
+	if blkdev.Index <= AcrnBlkDevPoolSz {
+		AcrnBlkdDevSlot[blkdev.Index] = slot
+	} else {
+		baselogger.Infof("Invalid index %d for the virtio-blk device!", blkdev.Index)
+	}
+
+	return acrnParams
+}
+
+// Valid returns true if the BridgeDevice structure is valid and complete.
+func (bridgeDev BridgeDevice) Valid() bool {
+	if bridgeDev.Function != 0 || bridgeDev.Emul != "hostbridge" {
+		return false
+	}
+	return true
+}
+
+// AcrnParams returns the acrn parameters built out of this bridge device.
+func (bridgeDev BridgeDevice) AcrnParams(slot int, config *Config) []string {
+	var acrnParams []string
+
+	acrnParams = append(acrnParams, "-s")
+	acrnParams = append(acrnParams, fmt.Sprintf("%d:%d,%s", slot,
+		bridgeDev.Function, bridgeDev.Emul))
+
+	return acrnParams
+}
+
+// Valid returns true if the BridgeDevice structure is valid and complete.
+func (lpcDev LPCDevice) Valid() bool {
+	return lpcDev.Emul == "lpc"
+}
+
+// AcrnParams returns the acrn parameters built out of this bridge device.
+func (lpcDev LPCDevice) AcrnParams(slot int, config *Config) []string {
+	var acrnParams []string
+	var deviceParams []string
+
+	acrnParams = append(acrnParams, "-s")
+	acrnParams = append(acrnParams, fmt.Sprintf("%d:%d,%s", slot,
+		lpcDev.Function, lpcDev.Emul))
+
+	//define UART port
+	deviceParams = append(deviceParams, "-l")
+	deviceParams = append(deviceParams, "com1,stdio")
+	acrnParams = append(acrnParams, strings.Join(deviceParams, ""))
+
+	return acrnParams
+}
+
+func (config *Config) appendName() {
+	if config.Name != "" {
+		config.acrnParams = append(config.acrnParams, config.Name)
+	}
+}
+
+func (config *Config) appendDevices() {
+	slot := 0
+	for _, d := range config.Devices {
+		if !d.Valid() {
+			continue
+		}
+
+		if slot == 2 {
+			slot++ /*Slot 2 is assigned for GVT-g in ACRN, so skip 2 */
+		}
+		config.acrnParams = append(config.acrnParams, d.AcrnParams(slot, config)...)
+		slot++
+	}
+}
+
+func (config *Config) appendUUID() {
+	if config.UUID != "" {
+		config.acrnParams = append(config.acrnParams, "-U")
+		config.acrnParams = append(config.acrnParams, config.UUID)
+	}
+}
+
+func (config *Config) appendACPI() {
+	if config.ACPIVirt {
+		config.acrnParams = append(config.acrnParams, "-A")
+	}
+}
+
+func (config *Config) appendMemory() {
+	if config.Memory.Size != "" {
+		config.acrnParams = append(config.acrnParams, "-m")
+		config.acrnParams = append(config.acrnParams, config.Memory.Size)
+	}
+}
+
+func (config *Config) appendCPUs() {
+	if config.NumCPU != 0 {
+		config.acrnParams = append(config.acrnParams, "-c")
+		config.acrnParams = append(config.acrnParams, fmt.Sprintf("%d", config.NumCPU))
+	}
+
+}
+
+func (config *Config) appendKernel() {
+	if config.Kernel.Path != "" {
```

Comment:
Personal preference, but I tend to invert logic like this to reduce the indent and make the code slightly clearer:

```go
if config.Kernel.Path == "" {
    return
}

config.acrnParams = ...
```

Reply:
done